lightning-c-bindings/demo.cpp

```diff
@@ -0,0 +1,480 @@
+
+	while (true) {
+		LDK::CVec_ChannelDetailsZ new_channels_2 = ChannelManager_list_channels(&cm2);
+		if (new_channels_2->datalen == 1) {
+			// Sample getting our counterparty's init features (which used to be hard to do without a memory leak):
+			const LDK::InitFeatures init_feats = ChannelDetails_get_counterparty_features(&new_channels_2->data[0]);
+			assert(init_feats->inner != NULL);
+			break;
+		}
+		std::this_thread::yield();
+	}
+
+	LDKEventsProvider ev1 = ChannelManager_as_EventsProvider(&cm1);
+	while (true) {
+		LDK::CVec_EventZ events = ev1.get_and_clear_pending_events(ev1.this_arg);
+		if (events->datalen == 1) {
+			assert(events->data[0].tag == LDKEvent_FundingGenerationReady);
+			assert(events->data[0].funding_generation_ready.user_channel_id == 42);
+			assert(events->data[0].funding_generation_ready.channel_value_satoshis == 40000);
+			assert(events->data[0].funding_generation_ready.output_script.datalen == 34);
+			assert(!memcmp(events->data[0].funding_generation_ready.output_script.data, channel_open_tx + 58, 34));
+			LDKThirtyTwoBytes txid;
+			for (int i = 0; i < 32; i++) { txid.data[i] = channel_open_txid[31-i]; }
+			LDK::OutPoint outp = OutPoint_new(txid, 0);
+			ChannelManager_funding_transaction_generated(&cm1, &events->data[0].funding_generation_ready.temporary_channel_id.data, outp);
+			break;
+		}
+		std::this_thread::yield();
+	}
+
+	// We observe when the funding signed messages have been exchanged by
+	// waiting for two monitors to be registered.
+	PeerManager_process_events(&net1);
+	while (true) {
+		LDK::CVec_EventZ events = ev1.get_and_clear_pending_events(ev1.this_arg);
+		if (events->datalen == 1) {
+			assert(events->data[0].tag == LDKEvent_FundingBroadcastSafe);
+			assert(events->data[0].funding_broadcast_safe.user_channel_id == 42);
+			break;
+		}
+		std::this_thread::yield();
+	}
+
+	BlockNotifier_block_connected(&blocks1, LDKu8slice { .data = channel_open_block, .datalen = sizeof(channel_open_block) }, 1);
+	BlockNotifier_block_connected(&blocks2, LDKu8slice { .data = channel_open_block, .datalen = sizeof(channel_open_block) }, 1);
+
+	BlockNotifier_block_connected(&blocks1, LDKu8slice { .data = block_1, .datalen = sizeof(block_1) }, 2);
+	BlockNotifier_block_connected(&blocks2, LDKu8slice { .data = block_1, .datalen = sizeof(block_1) }, 2);
+
+	BlockNotifier_block_connected(&blocks1, LDKu8slice { .data = block_2, .datalen = sizeof(block_2) }, 3);
+	BlockNotifier_block_connected(&blocks2, LDKu8slice { .data = block_2, .datalen = sizeof(block_2) }, 3);
+
+	PeerManager_process_events(&net1);
+	PeerManager_process_events(&net2);
+
+	// Now send funds from 1 to 2!
+	while (true) {
+		LDK::CVec_ChannelDetailsZ outbound_channels = ChannelManager_list_usable_channels(&cm1);
+		if (outbound_channels->datalen == 1) { break; }
+		std::this_thread::yield();
+	}
+
+	LDK::CVec_ChannelDetailsZ outbound_channels = ChannelManager_list_usable_channels(&cm1);
```

Comment:
Working on augmenting the demo (partly to get a better hands-on feel for the bindings). So far, I've added these lines here since presumably wallet devs might be interested in how to access channel data to display to users:

```
	// Example of accessing channel data
	const LDKChannelDetails *channel = &outbound_channels->data[0];
	const uint8_t (*channel_id)[32] = {ChannelDetails_get_channel_id(channel)};
	printf("New open channel! Channel details:\n");
	printf("Channel ID: ");
	for (int i = 0; i < 32; i++) {
		printf("%02x", *channel_id[i]);
	}
	printf("\n");
	LDKPublicKey remote_node_id = ChannelDetails_get_remote_network_id(channel);
	printf("Counterparty node ID: ");
	for (int i = 0; i < 33; i++) {
		printf("%02x", remote_node_id.compressed_form[i]);
	}
	printf("\n");
	uint64_t channel_value = ChannelDetails_get_channel_value_satoshis(channel);
	printf("Channel value (satoshis): %lu\n", channel_value);
	uint64_t outbound_capacity = ChannelDetails_get_outbound_capacity_msat(channel);
	printf("Outbound capacity (msat): %lu\n", outbound_capacity);
	uint64_t inbound_capacity = ChannelDetails_get_inbound_capacity_msat(channel);
	printf("Inbound capacity (msat): %lu\n", inbound_capacity);
	bool is_live = ChannelDetails_get_is_live(channel);
	printf("Ready to send payments: %s\n", is_live ? "true" : "false");
```
Other things possibly worth adding:
* channel close (working on this, still figuring out how to come up with blocks to confirm the close tx though)
* (de)serialization of channel data

Reply:
FWIW, the way I generated the build-in test case is by building a regular ol' rust-lightning test, printing the block/tx that was needed, and then panic!()ing the test.

Reply:
I added a few assertions for all the accessors you hit here.